rfd/0214-identifier-first-login.md

```diff
@@ -0,0 +1,193 @@
+automatically be shown the auth connector(s) relevant to them.
+
+## Details
+
+Identifier-first login will be configured in an auth connector's resource yaml. Any given auth
+connector can be configured with a `user_matchers` field containing a set of RegEx pattern(s) that
+can match usernames. When a user visits the login page, they will be prompted to provide a username,
+after which they'll be shown all the auth connectors configured to match it.
+
+If this field is not set, the auth connector will default to matching all users.
+
+It should be noted that this username does not represent anything beyond being a reference to map to
+connectors on the Teleport side, and users can enter any arbitrary string. For example, if a user
+enters `joe@foo.com` as their username and are mapped to an Okta connector, but their Okta user is
+actually `bob@foo.com`, they will be logged into Teleport as `bob@foo.com`. This detail should be
+explicitly mentioned in the docs.
+
+### Example matchers
+
+#### Match all usernames that end in `@foo.com`
+
+```yaml
+user_matchers: [".*@foo\.com$"]
+```
+
+### Match `joe@foo.com` and all usernames that end in `@goteleport.com`
+
+```yaml
+user_matchers: ["^joe@foo.com$", ".*@goteleport\.com$"]
+```
+
+### Example auth connector configuration:
+
+```yaml
+kind: saml
+metadata:
+  name: testconnector
+  revision: 60443273-15b2-4e5b-a2e2-9338439eac37
+spec:
+  acs: https://<cluster-url>/v1/webapi/saml/acs/testconnector
+  attributes_to_roles:
+    - name: groups
+      roles:
+        - editor
+      value: okta-admin
+    - name: groups
+      roles:
+        - auditor
+      value: okta-auditor
+    - name: groups
+      roles:
+        - access
+      value: okta-dev
+  audience: https://<cluster-url>/v1/webapi/saml/acs/testconnector
+  user_matchers: ["^joe@foo.com$", ".*@goteleport\.com$"]
+  cert: ''
+  display: Okta
+  entity_descriptor: |
+    <?xml version="1.0" encoding="UTF-8"?>
+    <md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" entityID="http://www.example.com/00000000000000000000">
+    ...
+    </md:EntityDescriptor>
+  entity_descriptor_url: ''
+  issuer: http://www.example.com/00000000000000000000
+  service_provider_issuer: https://teleport.com/v1/webapi/saml/acs/testconnector
+  signing_key_pair:
+    cert: |
+      -----BEGIN CERTIFICATE-----
+      ...
+      -----END CERTIFICATE-----
+    private_key: |
+      -----BEGIN RSA PRIVATE KEY-----
+      ...
+      -----END RSA PRIVATE KEY-----
+  sso: https://www.example.com/app/00000000000000000000/sso/saml
+version: v2
+```
+
+### Implementation
+
+Upon loading the Web UI, the server returns a web config which contains some details about the cluster
+including the Teleport edition, login methods and auth connectors (stored in `window['GRV_CONFIG']`).
+This will now also include the RegEx matchers per auth connector.
+
+All matching and filtering will be done on the client side. This is so that we can avoid having to
```

Comment:
Client side matching is efficient but it does mean the set of mappings is publicly available.

If we do the matching on the server side, we don't keep the mappings 100% confidential as you can make guesses against the unauthenticated API, but we're at least not sending the full list to anyone who asks for it.

Let's think about this a little more and decide whether it's a concern or not.

Reply:
This is a good point, I'll adjust the implementation so that the initial config response merely tells the UI whether or not identifier-first login is enabled so that it knows whether to display the "enter your email" prompt. Once the user submits their email, it'll hit an endpoint which does the actual matching and returns the relevant connectors. For remembered users we'll make that request as soon as the page is loaded and we detect that both 1) identifier-first login is enabled and 2) there is a remembered email stored.

I'll update the RFD with this change.